### PR TITLE
Added symfony validator feature

### PR DESCRIPTION
**Pull request by goetas**

### CODE REVIEW

File: src/Jms/YamlValidatorConverter.php [hunk 1]
```diff
@@ -117,47 +116,43 @@ private function loadValidatorType(array &$property, Type $type, $arrayized = fa
...
                         break;
                 }
-
-                if ($rule) {
-                    $rules[] = $rule;
-                }
             }
 
-            if ($rules) {
+            if (!!count($rules)) {
```
Comment (goetas) [inline src/Jms/YamlValidatorConverter.php]: please use a more intuitive boolean. `!==0` is more understandable

File: src/Jms/YamlValidatorConverter.php [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+                        foreach ($check as $item) {
+                            $rules[] = [
+                                'GreaterThanOrEqual' => $item['value']
+                            ];
+                        }
+                        break;
+                }
+            }
+
+            if (!!count($rules)) {
```
Comment (goetas) [inline src/Jms/YamlValidatorConverter.php]: please use `!==0` or `>0`

File: src/Jms/YamlValidatorConverter.php [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                }
+                $property["validator"][] = [
+                    'NotNull' => null
+                ];
+            }
+
+        }
+    }
+
+    protected function visitSimpleType(&$class, &$data, SimpleType $type, $name)
```
Comment (goetas) [inline src/Jms/YamlValidatorConverter.php:269]: we can have also arrays of objects, not just primitives, they have to be validated too.

essentially each property needs the rule `Valid` to apply recursive validation.

File: tests/Converter/JMS/Xsd2JmsBase.php [hunk 4]
```diff
@@ -38,7 +38,7 @@ protected function getClasses($xml)
     public function getBaseTypeConversions()
     {
         return [
-            ['xs:dateTime', 'DateTime'],
+            ['xs:dateTime', 'GoetasWebservices\\Xsd\\XsdToPhp\\XMLSchema\\DateTime'],
```
Comment (goetas) [inline tests/Converter/JMS/Xsd2JmsBase.php]: why this?

File: tests/Converter/Validator/Xsd2ValidatorBase.php [hunk 5]
```diff
@@ -0,0 +1,194 @@
+<?php
+namespace GoetasWebservices\Xsd\XsdToPhp\Tests\Converter\Validator;
+
+use GoetasWebservices\Xsd\XsdToPhp\Jms\YamlValidatorConverter;
+use GoetasWebservices\Xsd\XsdToPhp\Naming\ShortNamingStrategy;
+use GoetasWebservices\XML\XSDReader\SchemaReader;
+
+abstract class Xsd2ValidatorBase extends \PHPUnit_Framework_TestCase
```
Comment (goetas) [inline tests/Converter/Validator/Xsd2ValidatorBase.php]: currently there is only one class that inherits from this (`Xsd2ValidatorTest`) isn't this premature architecture?

File: tests/Converter/Validator/Xsd2ValidatorTest.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * @dataProvider getRestrictionsValidations
+     */
+    public function testComplexTypeWithValidations($xsRestrictions, $ymlValidations)
+    {
+        $content = '
+             <xs:schema targetNamespace="http://www.example.com" xmlns:xs="http://www.w3.org/2001/XMLSchema"  xmlns:ex="http://www.example.com">
+                <xs:complexType name="complexType-1">
+                    <xs:sequence>
+                        <xs:element name="column1">
```
Comment (goetas) [inline tests/Converter/Validator/Xsd2ValidatorTest.php:259]: this test is good, I will add another test, having `<xs:element name="column1" maxOccours="unbounded">`

File: tests/Converter/Validator/Xsd2ValidatorTest.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+{
+
+    /**
+     * @dataProvider getRestrictionsValidations
+     */
+    public function testSimpleTypeWithValidations($xsRestrictions, $ymlValidations)
+    {
+        $xml = "
+             <xs:schema targetNamespace=\"http://www.example.com\" xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">
+                <xs:element name=\"element-one\">
```
Comment (goetas) [inline tests/Converter/Validator/Xsd2ValidatorTest.php]: this test is good, I will add another test, having `<xs:element name="element-one" maxOccours="unbounded">`

File: tests/Converter/Validator/Xsd2ValidatorTest.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+        $this->assertEquals(
+            [
+                'Example\\ComplexType1Type' => [
+                    'properties' => [
+                        'column1' => array_merge(
+                            $ymlValidations,
+                            [
+                                [
+                                    'NotBlank' => null
```
Comment (goetas) [inline tests/Converter/Validator/Xsd2ValidatorTest.php]: pay attention, there is a subtle difference between `NotBlank` and `NotNull`.

In almost all the cases you should use not null probably

File: tests/Validator/ValidatorTest.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * @var ValidatorInterface
+     */
+    protected $validator;
+
+    public function setUp()
+    {
+        $loader = new ClassLoader();
+        $loader->addPsr4("OTA\\", "/home/goetas/projects/goetas-webservices-xsd2php/ota/php");
```
Comment (goetas) [inline tests/Validator/ValidatorTest.php]: this are just example paths.. they have to be configured.

File: src/Jms/YamlValidatorConverter.php [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+        // Required properties
+        if ($classType = $this->visitType($type)) {
+            if ($element->getMin() !== 0) {
+                if ($arrayized){
+                    $property["validator"][] = [
+                        'Count' => ['min' => 1]
+                    ];
+                }
+                $property["validator"][] = [
```
Comment (goetas) [inline src/Jms/YamlValidatorConverter.php]: probably this is not necessary (or is an constraint that we do not know yet if apply or not)

File: tests/Converter/JMS/Xsd2JmsBase.php (hunk 4 above)
Comment (davidsonalencar) [reply tests/Converter/JMS/Xsd2JmsBase.php]: This method is used as `dataProvider` in testElementOfBaseType. Adjusted to fill param $phpName with `GoetasWebservices\\Xsd\\XsdToPhp\\XMLSchema\\DateTime` value. See 32dc6ff

File: tests/Converter/Validator/Xsd2ValidatorTest.php (hunk 8 above)
Comment (davidsonalencar) [reply tests/Converter/Validator/Xsd2ValidatorTest.php]: Based [NotBlank doc](https://symfony.com/doc/current/reference/constraints/NotBlank.html) implement not equal to a blank string and also not equal to null. It's great for me. What do you think?

Comment (goetas) [reply tests/Converter/Validator/Xsd2ValidatorTest.php]: in xml, a node that says:

`minOccours=1, regex="[a-z]"`

and empty string is a valid value :)

File: src/Jms/YamlValidatorConverter.php (hunk 1 above)
Comment (davidsonalencar) [reply src/Jms/YamlValidatorConverter.php]: Applied better codding. See 3151dcb

File: src/Jms/YamlValidatorConverter.php (hunk 2 above)
Comment (davidsonalencar) [reply src/Jms/YamlValidatorConverter.php]: Applied better codding. See 3151dcb

File: tests/Converter/Validator/Xsd2ValidatorBase.php (hunk 5 above)
Comment (davidsonalencar) [reply tests/Converter/Validator/Xsd2ValidatorBase.php]: Fixed. See baff9be

File: tests/Converter/Validator/Xsd2ValidatorTest.php (hunk 8 above)
Comment (davidsonalencar) [reply tests/Converter/Validator/Xsd2ValidatorTest.php]: Perfect. You convinced me, :)

File: src/Jms/YamlValidatorConverter.php (hunk 10 above)
Comment (davidsonalencar) [reply src/Jms/YamlValidatorConverter.php]: At first is necessary for simple type/native type for indicate if required or not.

File: src/Jms/YamlValidatorConverter.php (hunk 3 above)
Comment (davidsonalencar) [reply src/Jms/YamlValidatorConverter.php:269]: Implemented feature to apply these rules. See c1699dc

File: tests/Converter/Validator/Xsd2ValidatorTest.php (hunk 6 above)
Comment (davidsonalencar) [reply tests/Converter/Validator/Xsd2ValidatorTest.php:259]: Created some tests in separate cases. See f42fe40

File: tests/Converter/Validator/Xsd2ValidatorTest.php (hunk 7 above)
Comment (davidsonalencar) [reply tests/Converter/Validator/Xsd2ValidatorTest.php]: Created some tests in separate cases. See f42fe40

**Comment by davidsonalencar** [comment]
Reading extended elements still need to be implemented.

**Comment by goetas** [comment]
If somebody is interested in sponsoring this feature I can continue to work on it.

Pricing can be agreed (compatible with west-european hourly rates).

**Comment by goetas** [comment]
After a lot of work, this is almost done. 
Validating the objects before or after xsd serialization/deserialization now is possible using the symfony/validator generated metadata!

Only a bit of documentation in missing.

**Comment by goetas** [comment]
Syntax should be:

```php
<?php

use Symfony\Component\Validator\Validation;

// get the validator
$builder = Validation::createValidatorBuilder();
foreach (glob('soap/validation/*.yml') as $file) {
    $builder->addYamlMapping($file);
}
$validator =  $builder->getValidator();

// validate $object
$violations = $validator->validate($object, null, ['xsd_rules']);

```

**Pull request merged**